airflow/jobs.py

```diff
@@ -415,13 +447,15 @@ def process_dag(self, dag, executor):
                     # in self.prioritize_queued
                     continue
                 else:
-                    # Trying to run the next schedule
-                    next_schedule = (
-                        ti.execution_date + task.schedule_interval)
-                    if (
-                            ti.task.end_date and
-                            next_schedule > ti.task.end_date):
+                    # Checking whether there is a DagRun for which a task
+                    # needs to be created
+                    qry = session.query(func.min(models.DagRun.execution_date)).filter(
+                        and_(models.DagRun.dag_id == dag.dag_id,
+                        models.DagRun.execution_date > ti.execution_date))
```

Comment:
If we have the state of the DAG run in `DagRun`, we could get a list of active runs outside of the tasks for loop, and attempt to schedule all tasks for all active job run.

This would solve confusing limitations documented on "The Scheduler" around never filling in holes and only moving forward from the last execution of each task.